src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -0,0 +1,110 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using FellowOakDicom;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Common;
+using Microsoft.Health.Dicom.Core.Features.Context;
+using Microsoft.Health.Dicom.Core.Features.Delete;
+using Microsoft.Health.Dicom.Core.Features.ExtendedQueryTag;
+using Microsoft.Health.Dicom.Core.Features.Model;
+
+namespace Microsoft.Health.Dicom.Core.Features.Workitem
+{
+    /// <summary>
+    /// Provides functionality to orchestrate the storing of the DICOM instance entry.
+    /// </summary>
+    public class WorkitemOrchestrator : IWorkitemOrchestrator
+    {
+        private readonly IDicomRequestContextAccessor _contextAccessor;
+        private readonly IIndexWorkitemStore _indexWorkitemStore;
+        private readonly IWorkitemStore _workitemStore;
+        private readonly IWorkitemQueryTagService _workitemQueryTagService;
+
+        public WorkitemOrchestrator(
+            IDicomRequestContextAccessor contextAccessor,
+            IWorkitemStore workitemStore,
+            IIndexWorkitemStore indexWorkitemStore,
+            IDeleteService deleteService,
+            IWorkitemQueryTagService workitemQueryTagService)
+        {
+            _contextAccessor = EnsureArg.IsNotNull(contextAccessor, nameof(contextAccessor));
+            _indexWorkitemStore = EnsureArg.IsNotNull(indexWorkitemStore, nameof(indexWorkitemStore));
+            _workitemStore = EnsureArg.IsNotNull(workitemStore, nameof(workitemStore));
+            _workitemQueryTagService = workitemQueryTagService;
+        }
+
+        /// <inheritdoc />
+        public async Task AddWorkitemAsync(
+            DicomDataset dataset,
+            CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+
+            WorkitemInstanceIdentifier identifier = null;
+
+            try
+            {
+                int partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
+
+                IReadOnlyCollection<QueryTag> queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
+
+                long workitemKey = await _indexWorkitemStore
+                    .AddWorkitemAsync(partitionKey, dataset, queryTags, cancellationToken)
+                    .ConfigureAwait(false);
+
+                // TODO: this needs to be done before storing the dataset in the Blob Store.
+                string workitemInstanceUid = dataset.GetString(DicomTag.SOPInstanceUID);
```

Comment:
This is not valid anymore. Either we should get from `AffectedSOPInstanceUID` or we do this while retrieving

Reply:
Lets do in retrieve or update response builder